instrumentation/opentelemetry-instrumentation-remoulade/src/opentelemetry/instrumentation/remoulade/__init__.py

```diff
@@ -0,0 +1,200 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Usage
+-----
+
+* Start broker backend
+
+::
+
+    docker run -p 5672:5672 rabbitmq
+
+* Run instrumented actor
+
+.. code-block:: python
+
+    from remoulade.brokers.rabbitmq import RabbitmqBroker
+    import remoulade
+
+    RemouladeInstrumentor().instrument()
+
+    broker = RabbitmqBroker()
+    remoulade.set_broker(broker)
+
+    @remoulade.actor
+    def multiply(x, y):
+        return x * y
+
+    broker.declare_actor(count_words)
+
+    multiply.send(43, 51)
+
+"""
+from typing import Collection, Iterable, List, Optional
+
+from remoulade import Middleware, broker
+
+from opentelemetry import trace
+from opentelemetry.instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry.instrumentation.remoulade import utils
+from opentelemetry.instrumentation.remoulade.package import _instruments
+from opentelemetry.instrumentation.remoulade.version import __version__
+from opentelemetry.propagate import extract, inject
+from opentelemetry.propagators.textmap import CarrierT, Getter
+from opentelemetry.semconv.trace import SpanAttributes
+
+_MESSAGE_TAG_KEY = "remoulade.action"
+_MESSAGE_SEND = "send"
+_MESSAGE_RUN = "run"
+
+_MESSAGE_NAME_KEY = "remoulade.actor_name"
+
+
+class RemouladeGetter(Getter):
+    def get(self, carrier: CarrierT, key: str) -> Optional[str]:
+        value = carrier.get(key, None)
+        if value is None:
+            return None
+        if isinstance(value, str) or not isinstance(value, Iterable):
+            value = (value,)
+        return value
+
+    def keys(self, carrier: CarrierT) -> List[str]:
+        return []
+
+
+remoulade_getter = RemouladeGetter()
+
+
+class InstrumentationMiddleware(Middleware):
+    def __init__(self, _tracer):
+        self._tracer = _tracer
+        self._span_registry = {}
+
+    def before_process_message(self, _broker, message):
+        if "trace_ctx" not in message.options:
+            return
+
+        trace_ctx = extract(
+            message.options["trace_ctx"], getter=remoulade_getter
+        )
+        retry_count = message.options.get("retries", None)
+
+        operation_name = (
+            "remoulade/process"
+            if retry_count is None
+            else f"remoulade/process(retry-{retry_count})"
+        )
+
+        span = self._tracer.start_span(
+            operation_name, kind=trace.SpanKind.CONSUMER, context=trace_ctx
+        )
+
+        if retry_count is not None:
```

Comment:
We want to be able to make sampling decisions based off of the attributes that are passed in. In this case, retry count should be added as an attribute and passed into `start_span`.